docs/design/agentcube-proposal.md

```diff
@@ -0,0 +1,475 @@
+2.  Secure, isolated code execution / code interpreter sessions
+    - Notebooks, “Run this code” buttons, and REPL-like experiences where user code should be executed in short-lived, isolated sandboxes.
+    - Per-session environments with strict resource limits, policy controls, and clear lifecycle management on Kubernetes.
+3.  Serverless-style agent orchestration
+    - Treating each agent invocation as a lightweight, event-driven workload, scheduled on demand and scaled to zero when idle.
+    - Integrating with existing Kubernetes primitives (autoscaling, scheduling, observability) while hiding complexity from application developers.
+4.  Multi-tenant AI platforms
+    - Platform teams providing standardized agent and interpreter runtimes to multiple product teams.
+    - Central management of runtime images, dependencies, security policies, and quota, while allowing tenants to customize tools and workflows.
+
+## Overall Architecture
+
+![](./images/agentcube.svg)
+
+### Concepts
+
+1.  **API: AgentRuntime**
+
+The CRD specifies the template to build agent runtime sandboxes and also can be used to specify the port, protocol, and url path that agent will be running with.
+
+2.  **API: CodeInterpreter**
+
+The CRD specifies the template to build code interpreter sandboxes and also can be used to specify the port, protocol, and url path that code interpreter will be running with.
+
+3.  **Component: Agentcube Router**
+
+A data plane component that used to do authentication, authorization, ratelimiting and routing for agent, codeinterpreter, MCP tools invocations.
+
+4.  **Component: Workload Manager**
+
+A control plane component that is responsible for sandbox creation/deletion, warmpool management and also idle sessions GC.
+
+5.  **Component: SharedSessionStorage(Redis)**
+
+A global shared session storage, which can be used to synchronize sessions across multiple replicas of **Agentcube Router Workload Manager** also exploits it to do sandboxes cleanup
+
+6.  **Session:**
+
+A session represents an abstract instance of an agent runtime or code interpreter. You can think of it as the lifetime of that agent or interpreter, during which it can handle one or more invocations.
+
+AgentCube isolates every session to enable secure, consistent reuse of context across multiple tool invocations. This isolation is critical for AI agent workloads, which often involve dynamic, multi-step execution.
+
+Each tool session runs inside its own microVM with dedicated CPU, memory, and filesystem resources. This design ensures that one user’s invocation cannot access data from another user’s session. When the session ends, the microVM is fully terminated and its memory is sanitized, eliminating the risk of cross-session data leakage.
+
+The session concept here is different from that from the view of agent application. From agent’s view, a session is the unit of conversational context, Within that session, the agent keeps state and memory so it can:
```

Comment:
```suggestion
The session concept here is different from that from the view of agent application. From agent’s view, a session is the unit of conversational context. Within that session, the agent keeps state and memory so it can:
```